server/summary_biplot_server.R

```diff
@@ -1,3 +1,295 @@
-output$SUMMARY_BIPLOT = renderPlot({
-    summary_biplot(1, 1)
+
+
+files <- site_data %>%
+    filter(site_type == "stream_gauge") %>%
+    select(domain, site_name) %>%
+    mutate(path_m = glue('data/biplot/month/{d}/{s}.feather',
+                         d = domain,
+                         s = site_name),
+           path_y = glue('data/biplot/year/{d}/{s}.feather',
+                         d = domain,
+                         s = site_name))
+
+observe({
+    doms <- input$DOMAINS2_S
+    
+    dom_sites <- site_data %>%
+        filter(site_type == 'stream_gauge') %>%
+        filter(domain %in%doms) %>%
+        pull(site_name)
+    
+    updateSelectInput(session, 'SITES2', choices = dom_sites)
+})
+
+observe({
+    data_type <- input$X_TYPE2
+    
+    if(data_type == 'Concentration') {
+        select <- chemvars_display_subset 
+        units <- conc_units}
+    
+    if(data_type == 'Flux') {
+        select <- fluxvars
+        units <- flux_units_bi }
+    
+    if(data_type == 'Discharge') {
+        select <- 'Q' 
+        units <- 'm^3' }
+    
+    updateSelectInput(session, 'X_VAR2', choices = select)
+    updateSelectInput(session, 'X_UNIT2', choices = units)
+    
+    })
+
+observe({
+    data_type <- input$Y_TYPE2
+    
+    if(data_type == 'Concentration') {
+        select <- chemvars_display_subset 
+        units <- conc_units}
+    
+    if(data_type == 'Flux') {
+        select <- fluxvars
+        units <- flux_units_bi }
+    
+    if(data_type == 'Discharge') {
+        select <- 'Q' 
+        units <- 'm^3' }
+    
+    updateSelectInput(session, 'Y_VAR2', choices = select)
+    updateSelectInput(session, 'Y_UNIT2', choices = units)
+    
+})
+
+observe({
+    data_type <- input$SIZE_TYPE2
+    
+    if(data_type == 'Concentration') {
+        select <- chemvars_display_subset 
+        units <- conc_units}
+    
+    if(data_type == 'Flux') {
+        select <- fluxvars
+        units <- flux_units_bi }
+    
+    if(data_type == 'Discharge') {
+        select <- 'Q' 
+        units <- 'm^3' }
+    
+    updateSelectInput(session, 'SIZE_VAR2', choices = select)
+    updateSelectInput(session, 'SIZE_UNIT2', choices = units)
+    
+})
+
+summary <- reactive({
+    
+    agg <- input$AGG2
+    
+    if(agg == 'YEARLY2') {
+        sum <- read_feather('data/biplot/year.feather')
+    }
+    
+    if(agg == 'MONTHLY2') {
+        sum <- read_feather('data/biplot/month.feather')
+    }
+    
+    return(sum)
+})
+
+filtered_bi <- reactive({
```

Comment:
in shiny, it's relatively easy to make reaction chains work like you want them to, but difficult to make them do so efficiently. one way is by preventing byproduct reactions that aren't relevant to the user's selections. within reactive() and observe() you can interrupt reactivity by wrapping inputs in isolate(). but an even clearer way to do this is with the alternatives to reactive() and observe(), which are eventReactive() and observeEvent(), respectively. the first argument to these functions determines which inputs can cause them to react.